src/util/unicode.cpp

```diff
@@ -224,43 +242,131 @@ Function: utf16_to_utf8
+    argv_narrow[i]=strdup(narrow(argv_wide[i]).c_str());
 
-  return result;
+  return argv_narrow;
 }
 
 /*******************************************************************\
 
-Function: narrow_argv
+Function: swap_last_bytes
 
-  Inputs:
+  Inputs: An integer
 
- Outputs:
+ Outputs: An integer whose representation swaps the last two bytes
 
- Purpose:
+ Purpose: A helper function for dealing with different UTF16  endians
 
 \*******************************************************************/
 
-const char **narrow_argv(int argc, const wchar_t **argv_wide)
+unsigned int swap_last_bytes(unsigned int x)
 {
-  if(argv_wide==NULL)
-    return NULL;
+    unsigned int b1=x & 0xFF;
+    unsigned int b2=x & 0xFF00;
+    return (b1 << 8)+(b2 >> 8);
+}
 
-  // the following never gets deleted
-  const char **argv_narrow=new const char *[argc+1];
-  argv_narrow[argc]=0;
+void utf16_append_code(unsigned int code, bool swap_bytes, std::wstring &result)
+{
+  if(code<0xFFFF)
+  { // code is encoded as one UTF16 character
+    // we just take the code and possibly swap the bytes
+    unsigned int a=(swap_bytes)?swap_last_bytes(code):code;
+    result+=static_cast<wchar_t>(a);
+  }
+  else // code is endoded as two UTF16 characters
+  {
+    // if this is valid unicode, we have
+    // code<0x10FFFF
+    // but let's not check it programmatically
+
+    // encode the code in UTF16, possibly swapping bytes.
+    code=code-0x10000;
+    unsigned int i1=((code>>10) & 0x3ff) | 0xD800;
+    unsigned int a1=(swap_bytes)?swap_last_bytes(i1):i1;
+    result+=static_cast<wchar_t>(a1);
+    unsigned int i2=(code & 0x3ff) | 0xDC00;
+    unsigned int a2=(swap_bytes)?swap_last_bytes(i2):i2;
+    result+=static_cast<wchar_t>(a2);
+  }
+}
 
-  for(int i=0; i<argc; i++)
-    argv_narrow[i]=strdup(narrow(argv_wide[i]).c_str());
 
-  return argv_narrow;
+/*******************************************************************\
+
+Function: utf8_to_utf16
+
+  Inputs: String in UTF-8 format, bool value indicating whether the
+          endianness should be different from the architecture one.
+
+ Outputs: String in UTF-16 format. The encoding follows the
+          endianness of the architecture iff swap_bytes is true.
+
+ Purpose:
+
+\*******************************************************************/
+std::wstring utf8_to_utf16(const std::string& in, bool swap_bytes)
+{
+    std::wstring result;
+    result.reserve(in.size());
+    int i=0;
+    while(i<in.size())
+    {
+      unsigned char c=in[i++];
+      unsigned int code=0;
+      // the ifs that follow find out how many UTF8 characters (1-4) store the
+      // next unicode character. This is determined by the few most
+      // significant bits.
+      if(c<=0x7F)
+      {
+        // if it's one character, then code is exactly the value
+        code=c;
+      }
+      else if(c<=0xDF)
+      { // in other cases, we need to read the right number of chars and decode
+        code=(c & 0x1F) << 6;
+        c=in[i++];
```

Comment:
Should check array bounds when trying to continue reading a code-point, in case the input is not really UTF-8